### ISSUE TITLE
CCSDTQ in parallel

### ISSUE THREAD

**Issue by jgonthier** [nwchemgit/nwchem#105]
Hello,

It seems I cannot make CCSDTQ work in parallel. On a simple input file with a single C atom:
```
memory stack 500 mb heap 500 mb global 2000 mb
geometry  units angstrom
  symmetry c1
  C 0.0 0.0 0.0
end

basis "ao basis" spherical
  * library cc-pVTZ
end
set lindep:tol 1e-10
SCF
  thresh 1e-6
  nopen 2
  UHF
end
tce
  SCF
  CCSDTQ
  thresh 1e-6
  io ga
  freeze atomic
end
task scf energy
task tce energy
```
I can run the job serially, it takes about 10-15 min with 40s per CCSDTQ iteration. When I try to run it in parallel using ```mpirun -np 2 nwchem input >& output &```, the job prints the header for the CCSDTQ iterations (after the integral transform), but then hangs while seemingly using both CPUs. Even after 20-30min the first iteration is not printed.

I compiled this version of NWChem with gfortran/gcc 7.3.1, openblas-serial64 and the following environment variables:
```
export NWCHEM_TOP=/root/nwchem-6.8.1
export USE_MPI=y
export NWCHEM_TARGET=LINUX64
export USE_PYTHONCONFIG=y
export PYTHONVERSION=2.7
export PYTHONHOME=/usr
export BLASOPT="-lnwclapack -lopenblas64"
export BLAS_SIZE=8
export PATH=/usr/lib64/openmpi/bin/:$PATH
export LD_LIBRARY_PATH=/usr/lib64/openmpi/lib/:$LD_LIBRARY_PATH
export USE_ARUR=y
export CCSDTQ="yes"
make nwchem_config NWCHEM_MODULES="all python"
```
Is there some option I am forgetting for CCSDTQ? The same code runs CCSDT in parallel with the expected speed-up.

**Comment by jeffhammond** [comment]
This is odd. Please include complete output file and a back trace to reveal where the code is hung (‘gstack pid’ does this on most Linux systems).

**Comment by jgonthier** [comment]
Ok, additional info: I'm running on a Docker container, so it starts from a very barebone Fedora 26, and then I install everything. It means something could potentially be missing.
Here is the output file:
[output.txt](https://github.com/nwchemgit/nwchem/files/2931732/output.txt)
gstack for the first NWChem process:
```
Thread 3 (Thread 0x7f7bd6f9d700 (LWP 79)):
#0  0x00007f7bd8e98a5d in poll () from /lib64/libc.so.6
#1  0x00007f7bd88c6253 in poll_dispatch () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f7bd88bd46b in opal_libevent2022_event_base_loop () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#3  0x00007f7bd8884c9e in progress_engine () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#4  0x00007f7bda72a36d in start_thread () from /lib64/libpthread.so.0
#5  0x00007f7bd8ea4b4f in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f7bd779e700 (LWP 77)):
#0  0x00007f7bd8ea50b3 in epoll_wait () from /lib64/libc.so.6
#1  0x00007f7bd88b93cb in epoll_dispatch () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f7bd88bd46b in opal_libevent2022_event_base_loop () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#3  0x00007f7bd77c5c4d in progress_engine () from /usr/lib64/openmpi/lib/openmpi/mca_pmix_pmix112.so
#4  0x00007f7bda72a36d in start_thread () from /lib64/libpthread.so.0
#5  0x00007f7bd8ea4b4f in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f7bddaf2740 (LWP 75)):
#0  0x00007f7bd88d8a43 in opal_timer_base_get_usec_sys_timer () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#1  0x00007f7bd8880509 in opal_progress () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f7bdab9259d in ompi_request_default_wait () from /usr/lib64/openmpi/lib/libmpi.so.20
#3  0x00007f7bdabf480a in ompi_coll_base_barrier_intra_two_procs () from /usr/lib64/openmpi/lib/libmpi.so.20
#4  0x00007f7bdaba6c6a in PMPI_Barrier () from /usr/lib64/openmpi/lib/libmpi.so.20
#5  0x0000000003ea17b6 in armci_tcgmsg_nxtval ()
#6  0x0000000003ea053a in nxtval_ ()
#7  0x000000000270b425 in ccsdtq_e_1_2_ ()
#8  0x000000000270c4b1 in ccsdtq_e_ ()
#9  0x000000000173d269 in tce_energy (rtdb=0, excitedfragment=.TRUE.) at tce_energy.F:4405
#10 0x0000000001725f0d in tce_energy_fragment (rtdb=0) at tce_energy_fragment.F:88
#11 0x00000000005447d7 in task_energy_doit (rtdb=0, theory=<error reading variable: Cannot access memory at address 0x7fff9cf4d000>, energy=6.0134700169990685e-154, _theory=_theory@entry=32) at task_energy.F:299
#12 0x0000000000545e9d in task_energy (rtdb=0) at task_energy.F:122
#13 0x0000000000538b45 in task (rtdb=0) at task.F:371
#14 0x00000000005310f8 in nwchem () at nwchem.F:313
#15 0x00000000005316be in main (argc=<optimized out>, argv=<optimized out>) at nwchem.F:397
#16 0x00007f7bd8db488a in __libc_start_main () from /lib64/libc.so.6
#17 0x000000000052f6ca in _start ()
```
and for the second process (somehow it's a bit different):
```
Thread 3 (Thread 0x7f9220b78700 (LWP 80)):
#0  0x00007f9222a73a5d in poll () from /lib64/libc.so.6
#1  0x00007f92224a1253 in poll_dispatch () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f922249846b in opal_libevent2022_event_base_loop () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#3  0x00007f922245fc9e in progress_engine () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#4  0x00007f922430536d in start_thread () from /lib64/libpthread.so.0
#5  0x00007f9222a7fb4f in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f9221379700 (LWP 78)):
#0  0x00007f9222a800b3 in epoll_wait () from /lib64/libc.so.6
#1  0x00007f92224943cb in epoll_dispatch () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f922249846b in opal_libevent2022_event_base_loop () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#3  0x00007f92213a0c4d in progress_engine () from /usr/lib64/openmpi/lib/openmpi/mca_pmix_pmix112.so
#4  0x00007f922430536d in start_thread () from /lib64/libpthread.so.0
#5  0x00007f9222a7fb4f in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f92276cd740 (LWP 76)):
#0  0x00007f92224b3a43 in opal_timer_base_get_usec_sys_timer () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#1  0x00007f922245b509 in opal_progress () from /usr/lib64/openmpi/lib/libopen-pal.so.20
#2  0x00007f9218eaf093 in mca_pml_ob1_iprobe () from /usr/lib64/openmpi/lib/openmpi/mca_pml_ob1.so
#3  0x00007f9224794b1e in PMPI_Iprobe () from /usr/lib64/openmpi/lib/libmpi.so.20
#4  0x0000000003ee69f2 in comex_make_progress () at ../../ga-5.6.5/comex/src-mpi/comex.c:337
#5  0x0000000003ee7c15 in comex_wait_all (group=<optimized out>) at ../../ga-5.6.5/comex/src-mpi/comex.c:1522
#6  comex_wait_proc (group=<optimized out>, proc=0) at ../../ga-5.6.5/comex/src-mpi/comex.c:1500
#7  comex_gets (src_ptr=0x50685e88, src_stride_ar=0x7ffd35ad9d74, dst_ptr=0x4f9968c0, dst_stride_ar=0x7ffd35ad9d54, count=0x7ffd35ad9d34, stride_levels=1, proc=0, group=<optimized out>) at ../../ga-5.6.5/comex/src-mpi/comex.c:914
#8  0x0000000003ee8fb5 in comex_nbgets (src_ptr=<optimized out>, src_stride_ar=<optimized out>, dst_ptr=<optimized out>, dst_stride_ar=<optimized out>, count=<optimized out>, stride_levels=<optimized out>, proc=0, group=0, hdl=0x7ffd35ad9620) at ../../ga-5.6.5/comex/src-mpi/comex.c:1550
#9  0x0000000003ee41b4 in PARMCI_NbGetS (src_ptr=<optimized out>, src_stride_arr=<optimized out>, dst_ptr=<optimized out>, dst_stride_arr=<optimized out>, count=<optimized out>, stride_levels=<optimized out>, proc=0, nb_handle=0x7ffd35ad9620) at ../../ga-5.6.5/comex/src-armci/armci.c:401
#10 0x0000000003e31e87 in ngai_nbgets ()
#11 0x0000000003e320d0 in ngai_gets ()
#12 0x0000000003e3a395 in ngai_get_common ()
#13 0x0000000003e3cd2b in wnga_get ()
#14 0x0000000003dd183c in ga_get_ ()
#15 0x00000000017757dc in get_block_ ()
#16 0x00000000017a2a04 in get_hash_block_ ()
#17 0x000000000270b866 in ccsdtq_e_1_2_ ()
#18 0x000000000270c4b1 in ccsdtq_e_ ()
#19 0x000000000173d269 in tce_energy (rtdb=0, excitedfragment=.TRUE.) at tce_energy.F:4405
#20 0x0000000001725f0d in tce_energy_fragment (rtdb=0) at tce_energy_fragment.F:88
#21 0x00000000005447d7 in task_energy_doit (rtdb=0, theory=<error reading variable: Cannot access memory at address 0x7ffd35ae0000>, energy=6.0134700169990685e-154, _theory=_theory@entry=32) at task_energy.F:299
#22 0x0000000000545e9d in task_energy (rtdb=0) at task_energy.F:122
#23 0x0000000000538b45 in task (rtdb=0) at task.F:371
#24 0x00000000005310f8 in nwchem () at nwchem.F:313
#25 0x00000000005316be in main (argc=<optimized out>, argv=<optimized out>) at nwchem.F:397
#26 0x00007f922298f88a in __libc_start_main () from /lib64/libc.so.6
#27 0x000000000052f6ca in _start ()
```

**Comment by jeffhammond** [comment]
My suspicion is that it isn't stuck, it's just hideously slow. Are you running in a VM?  As you can see from the backtrace, you have 3 threads per process, so if you are in a cloud environment with 2 virtual cores, you are oversubscribing badly, given that the threads are polling.

What ARMCI/ComEx conduit are you using?  I can see it is using MPI but can't tell if it is MPI-PR, MPI-TS, MPI-SPAWN, or MPI-MT.

**Comment by edoapra** [comment]
@jeffhammond I agree with your conclusion.
It is likely that ARMCI_NETWORK was not defined, therefore MPI-TS was chosen. Have a look at this line of the debug stack
```
#4  0x0000000003ee69f2 in comex_make_progress () at ../../ga-5.6.5/comex/src-mpi/comex.c:337
```
I would make the same suggestion to switch to MPI-PR .
ARMCI_NETWORK=SOCKETS might be better, too
 https://github.com/nwchemgit/nwchem-dockerfiles/blob/master/nwchem-681.fedora.sockets/Dockerfile
https://hub.docker.com/r/edoapra/nwchem-681.fedora.sockets

**Comment by jgonthier** [comment]
So I am running the docker image directly on my Mac, which has 6 cores. I let Docker use half of them. I am honestly not sure which ARMCI/ComEx conduit I am using. I followed the instructions at https://github.com/nwchemgit/nwchem/wiki/Compiling-NWChem#nwchem-66-on-fedora-22 and I didn't set any additional environment variables. Is there a way to extract this information from my NWChem installation?

Also, I have another installation that is directly on my Mac (using NWChem internal BLAS) and it exhibits the same behaviour.

**Comment by jeffhammond** [comment]
I don’t use Docker but I’ll test on my Mac later. At least with ARMCI-MPI, communication should go through shared memory and not require any network polling.

**Comment by jgonthier** [comment]
Great news: I built a docker image using the Dockerfile provided by @edoapra, simply adding the CCSDTQ=yes environment variable, and I could run CCSDTQ on 2 processes, and I do see a speed-up of about 1.7x. I should be able to work from that to get what I need running smoothly.
Closing the issue for now, I may post another comment later to confirm the problem is solved in my original image by just setting ARMCI_NETWORK appropriately.

Thank you very much for your help!
